### ISSUE TITLE
`no-useless-spread` False positive with iterator helpers

### ISSUE THREAD

**Issue by RebeccaStevens** [sindresorhus/eslint-plugin-unicorn#2018]
When using a polyfill for [TC39 Iterator Helpers](https://github.com/tc39/proposal-iterator-helpers), trying to convert a generator produced by one of the helpers to an array via the spread operator, results in said spread operator being marked as useless.

Rule effected: `unicorn/no-useless-spread`

Example:

```js
const data = new Map([["a", 1], ["b", 2], ["c", 3]]);
const foo = [...data.values().map((d) => d * 2)]; // [2, 4, 6]
```

is incorrectly flagged and fixed to:

```js
const data = new Map([["a", 1], ["b", 2], ["c", 3]]);
const foo = data.values().map((d) => d * 2); // Generator<number, void, undefined>
```

**Comment by AlexXanderGrib** [comment]
Also false positive with strings

```typescript
/**
 *
 *
 * @export
 * @param {string} base64
 * @return {Uint8Array} Uint8Array
 */
export function base64ToUint8Array(base64: string): Uint8Array {
  const alphabet =
    "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/".split(""); // base64 alphabet
  const result: number[] = [];

  for (let index = 0; index < base64.length / 4; index++) {
    const chunk = [...base64.slice(4 * index, 4 * index + 4)];
    //             ^^^
    // Unnecessarily cloning an array. eslint(unicorn/no-useless-spread)
    const bin = chunk
      .map((x) => alphabet.indexOf(x).toString(2).padStart(6, "0"))
      .join("");
    const bytes = bin.match(/.{1,8}/g)?.map((x) => Number.parseInt(x, 2)) ?? [];
    result.push(
      ...bytes.slice(
        0,
        3 -
          Number(getByIndex(base64, 4 * index + 2) === "=") -
          Number(getByIndex(base64, 4 * index + 3) === "=")
      )
    );
  }
  return new Uint8Array(result);
}
```

**Comment by domoritz** [comment]
Also false positives with typed arrays that one may want to be converted to an array of numbers. 

```ts
[...int32Array.slice(1)]
```

**Comment by fregante** [comment]
@AlexXanderGrib duplicate of https://github.com/sindresorhus/eslint-plugin-unicorn/issues/1988

@domoritz similar to that, but the same concept

**Comment by RebeccaStevens** [comment]
These sorts of false positives seem to be quite a common occurrence in this plugin, effecting multiple rules. I think it would be a good idea to try and come up with a more general solution than a hotfix.

If a user is using TypeScript, it would be best to utilize type information to determine if the value's type is of the desired type for the rule to apply to. In this case, that the value is an array. This would be a much more robust approach than simply looking at the name of the method alone.

This is the approach we take over at [eslint-functional](https://github.com/eslint-functional/eslint-plugin-functional). We have config options for JS users to configure whether types should be assumed (like what this plugin is currently doing) or not. [Here's an example of how we use this](https://github.com/eslint-functional/eslint-plugin-functional/blob/main/src/util/typeguard.ts#L386-L398).

**Comment by fregante** [comment]
Yes that’s also an issue with HTML live collections like .children, for example looping .children and altering the parent might enter an infinite loop
